hearth/lib/hearth/connection_pool.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+module Hearth
+  # @api private
+  class ConnectionPool
+    @pools_mutex = Mutex.new
+    @pools = {}
+
+    class << self
+      # @return [ConnectionPool]
+      def for(config = {})
+        @pools_mutex.synchronize do
+          puts "using connection pool for #{config}\n" if @pools[config]
+          puts "creating connection pool for #{config}\n" unless @pools[config]
+          @pools[config] ||= new
+        end
+      end
+
+      # @return [Array<ConnectionPool>] Returns a list of the
+      #   constructed connection pools.
+      def pools
+        @pools_mutex.synchronize do
+          @pools.values
+        end
+      end
+    end
+
+    # @api private
+    def initialize
+      @pool_mutex = Mutex.new
+      @pool = {}
+    end
+
+    # @param [URI::HTTP, URI::HTTPS] endpoint The HTTP(S) endpoint
+    #    to connect to (e.g. 'https://domain.com').
+    # @param [Proc] block A block that returns a new session.
+    # @return [Object, nil]
+    def session_for(endpoint, &block)
+      session = nil
+      endpoint = remove_path_and_query(endpoint)
+      # attempt to recycle an already open session
+      @pool_mutex.synchronize do
+        _clean
+        session = @pool[endpoint].shift if @pool.key?(endpoint)
+        puts "using session from pool: #{session}\n" if session
+      end
+      session || block.call
```

Comment:
Is there anything that keeps this from growing in an unbounded way?  (ie, a max connections or something like that).  Kicking off a ton of concurrent requests will create a ton of connections, which could be expensive

Reply:
It's bounded by uniqueness of configuration and uniqueness of endpoint. I think practically speaking, it will not grow large, and any new request will clean old connections. The way the code is currently, a ton of concurrent requests to the same endpoint/service and with the same configuration will only have one connection which is re-used.

Reply:
I believe a large number of concurrent requests will each create a new endpoint.

And the number of endpoints may be very large - consider virtual hosting of s3 buckets - each bucket will end up with its own pool of connections.

Reply:
Talked offline, it's not an issue in v3, inclined to leave it out for now.